core/src/main/java/com/linecorp/armeria/client/DefaultClientRequestContext.java

```diff
@@ -91,13 +94,19 @@ public DefaultClientRequestContext(
 
         final HttpHeaders headers = options.getOrElse(ClientOption.HTTP_HEADERS, HttpHeaders.EMPTY_HEADERS);
         if (!headers.isEmpty()) {
-            final HttpHeaders headersCopy = new DefaultHttpHeaders(true, headers.size());
-            headersCopy.set(headers);
-            attr(HTTP_HEADERS).set(headersCopy);
+            createAdditionalHeadersIfAbsent();
+            additionalRequestHeaders.set(headers);
         }
+
         runThreadLocalContextCustomizer();
     }
 
+    private void createAdditionalHeadersIfAbsent() {
```

Comment:
Because `additionalRequestHeaders` is volatile, we need to reduce the number of reads:

```java
private HttpHeaders createAdditionalHeadersIfAbsent() {
    final HttpHeaders additionalRequestHeaders = this.additionalRequestHeaders;
    if (additionalRequestHeaders == null) {
        return this.additionalRequestHeaders = new DefaultHttpHeaders();
    } else {
        return additionalRequestHeaders;
    }
}
```

... and see below.